frappe/deferred_insert.py

```diff
@@ -27,28 +30,34 @@ def deferred_insert(doctype: str, records: list[Union[dict, "Document"]] | str):
 
 def save_to_db():
 	queue_keys = frappe.cache.get_keys(queue_prefix)
+	# TODO: can this be async?
 	for key in queue_keys:
-		record_count = 0
 		queue_key = get_key_name(key)
 		doctype = get_doctype_name(key)
-		while frappe.cache.llen(queue_key) > 0 and record_count <= 500:
-			records = frappe.cache.lpop(queue_key)
-			records = json.loads(records.decode("utf-8"))
-			if isinstance(records, dict):
-				record_count += 1
-				insert_record(records, doctype)
-				continue
-			for record in records:
-				record_count += 1
-				insert_record(record, doctype)
+		while cache_size := frappe.cache.llen(queue_key):
+			records = frappe.cache.lpop(queue_key, count=cache_size)
+			records: TYPE_UNION_DICT_DOCUMENT = json.loads(records.decode("utf-8"))
+			if failed_records := insert_records(records, doctype):
+				frappe.cache.rpush(queue_key, json.dumps(failed_records))
```

Comment:
assumption is that failed records due to exceptions is an edge case and this wouldn't cause an infinite loop, in which case a max retry cap could be added which defeats the point of the issue.